### PR TITLE
added code to publish ipa to itunes

### PR DESCRIPTION
**Pull request by sierrakaplan**
# Why

Allows a snack ios app to also be pushed to itunes Connect using fastlane. Adds job config options to make this possible. See: expo/universe#3546

# How

Added a function that is sometimes called after building the ipa, to upload that ipa to iTunes Connect using fastlane.

# Test Plan

### CODE REVIEW

File: src/job.ts [hunk 1]
```diff
@@ -12,6 +12,12 @@ export interface IJob {
     // ios
     buildType?: IOS.BUILD_TYPES;
     bundleIdentifier?: string;
+    upload?: boolean;
+    options?: {
```
Comment (dsokal) [inline src/job.ts]: Please name the key more descriptive, so it indicates that these are Apple Credentials.

File: src/builders/utils/uploader.ts [hunk 2]
```diff
@@ -1,15 +1,23 @@
 import * as fs from 'fs-extra';
 
 import { uploadFile } from 'turtle/aws/s3';
+import { IContext } from 'turtle/builders/ios/context';
 import config from 'turtle/config';
 import logger from 'turtle/logger';
 
+import { ExponentTools } from 'xdl';
```
Comment (dsokal) [inline src/builders/utils/uploader.ts]: We try to follow this convention of imports:
- import native node modules first (and add an empty line if any)
- import any package from npm (and add an empty line if any)
- import any module from your project

File: src/builders/utils/uploader.ts [hunk 3]
```diff
@@ -29,3 +37,37 @@ export async function uploadBuildToS3(ctx: IUploadCtx) {
...
+    FASTLANE_DISABLE_COLORS: 1,
+    CI: 1,
+    LC_ALL: 'en_US.UTF-8',
+    FASTLANE_PASSWORD: options.password,
+  };
+
+  const fastlaneArgs = [
+    'run upload_to_testflight',
+    'username:',
+    '"' + options.username + '"',
```
Comment (dsokal) [inline src/builders/utils/uploader.ts]:
```suggestion
    `"${options.username}"`,
```
Are double quotes necessary here?

File: src/builders/utils/uploader.ts [hunk 4]
```diff
@@ -29,3 +37,37 @@ export async function uploadBuildToS3(ctx: IUploadCtx) {
...
+    'ipa:',
+    ctx.uploadPath,
+    'apple_id:',
+    ctx.appUUID,
```
Comment (dsokal) [inline src/builders/utils/uploader.ts]: It's not an Apple ID. `ctx.appUUID` is a randomly generated uuid.

File: src/builders/utils/uploader.ts (hunk 3 above)
Comment (sierrakaplan) [reply src/builders/utils/uploader.ts]: (fixed)  I was using the CLI documentation for running the command.

File: src/builders/utils/uploader.ts [hunk 5]
```diff
@@ -29,3 +39,62 @@ export async function uploadBuildToS3(ctx: IUploadCtx) {
...
+  const produceArgs = [
+    'produce',
+    '--username',
+    options.appleId,
+    '--app_identifier', // bundle id
+    bundleIdentifier,
+    '--app_name',
+    options.appName,
+    '--app_version',
+    '1.0',
```
Comment (dsokal) [inline src/builders/utils/uploader.ts]: Why `1.0`? Is this upload feature meant to only upload one version of an application to Testflight? If no then what's the way of making sure a user builds and uploads the app only once?

File: src/builders/utils/uploader.ts [hunk 6]
```diff
@@ -29,3 +39,62 @@ export async function uploadBuildToS3(ctx: IUploadCtx) {
...
+
+  const uploadArgs = [
+    'pilot',
+    'upload',
+    '--username',
+    options.appleId,
+    '--ipa',
+    ctx.uploadPath,
+    '--apple_id', // The unique App ID provided by App Store Connect
+    process.env.PRODUCE_APPLE_ID,
```
Comment (dsokal) [inline src/builders/utils/uploader.ts]: How does it end up set to an env variable?

File: src/job.ts [hunk 7]
```diff
@@ -12,6 +12,14 @@ export interface IJob {
...
```
Comment (dsokal) [inline src/job.ts]:
```suggestion
    uploadOptions?: {
```

File: src/builders/utils/uploader.ts [hunk 8]
```diff
@@ -29,3 +39,62 @@ export async function uploadBuildToS3(ctx: IUploadCtx) {
...
+
+export async function uploadBuildToTestFlight(ctx: IContext, options: IJobOptions, bundleIdentifier?: string) {
+  const fastlaneEnvVars = {
+    FASTLANE_USER: options.appleId,
+    FASTLANE_SKIP_UPDATE_CHECK: 1,
+    FASTLANE_DISABLE_COLORS: 1,
+    CI: 1,
+    LC_ALL: 'en_US.UTF-8',
+    FASTLANE_PASSWORD: options.applePassword,
+    FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD: options.appSpecificPassword,
```
Comment (dsokal) [inline src/builders/utils/uploader.ts]: Does it make sense to pass both `FASTLANE_PASSWORD` and `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD` as env vars? If not then we need to add some condition here.

**Comment by dsokal** [comment]
Please make sure the code you submit to review compiles and that you tested it at least once.

**Comment by zetaron** [comment]
@sierrakaplan I think the link to expo/universe is broken GitHub gives me 404s when I try to open it.

**Comment by dsokal** [comment]
@zetaron it's a private repository :)

**Pull request closed without merging**